### PR TITLE
Feature/wjiang/vertical indices

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
## Description

This PR tries to improve the performance of vertical interpolation by saving the indexes and the linear weight.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
@bena-nasa , at this point, only the case ( size(ptrin,3) == size(this%pl3d,3) ) is implemented.  But the results are not correct. Do I miss any step?

**Comment by weiyuan-jiang** [comment]
> @weiyuan-jiang You might need to do the conflict resolution on the multigroupserver. I took a look and I'm not sure which ones are right.

I messed the local and remote. I will fix it.

File: CHANGELOG.md [hunk 1]
```diff
@@ -53,6 +54,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Checked if a file is being written before sending out request to a writer
```
Comment (mathomp4) [inline CHANGELOG.md:57]:
```suggestion
```
I think this can be removed since I think this was the code that is in 2.39.7?

**Comment by weiyuan-jiang** [comment]
@tclune @bena-nasa A temporary test shows the improvement of one output  on collection of geosgcm_prog. ( those timer will be removed before merging).  How do we get total time for all ranks for that timer?
...
...
 new_way_total:   7.074475288391113E-004
 old_way_total:   7.667481899261475E-003
 new_way_total:   6.859898567199707E-004
 old_way_total:   7.682621479034424E-003
 new_way_total:   6.768256425857544E-004
 old_way_total:   7.748976349830627E-003
 new_way_total:   6.947368383407593E-004
 old_way_total:   7.644027471542358E-003
 new_way_total:   7.014274597167969E-004
 old_way_total:   7.637321949005127E-003
 new_way_total:   7.352679967880249E-004
 old_way_total:   7.577374577522278E-003
 new_way_total:   7.223784923553467E-004
old_way_total:   7.627740502357483E-003
new_way_total:   7.190555334091187E-004
old_way_total:   7.645726203918457E-003
...
...

**Comment by tclune** [comment]
(1) I don't think you want the total.   You might want max.

(2) The MPI support is all through DistributedProfiler, but you are using a simple timer.    For now you'll have to do MPI reduce (sum/max/...) calls directly with the relevant communicator.

**Comment by weiyuan-jiang** [comment]
Ah, I am using my eyes and find out the max. There are only 9 variables with vertical dimension in that collection ( excluding ZLE)
new_way_total:   7.738322019577026E-004
old_way_total:   1.003018021583557E-002

**Comment by weiyuan-jiang** [comment]
More realistic c360 run.  max  

 new_way_total:   2.666312456130981E-002
 old_way_total:   0.348897576332092

**Comment by weiyuan-jiang** [comment]
This PR now is zero-diff with the PR #2251

File: base/MAPL_VerticalMethods.F90 [hunk 2]
```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
...
+        levo = size(ptrout,3)
+        ptrout = MAPL_UNDEF
+        do lev = 1, levo
+          do j = 1, D2
+            do i = 1, D1
+               k = ks_(i,j,lev)
+               if (k == -1) cycle
+               !if (k == km) then
+               !   ptrout(i,j,lev) = ptrin(i,j,k)
+               !   cycle
```
Comment (tclune) [inline base/MAPL_VerticalMethods.F90:365]: Can we delete the commented code?
```suggestion
               !if (k == km) then
               !   ptrout(i,j,lev) = ptrin(i,j,k)
               !   cycle
```

File: base/MAPL_VerticalMethods.F90 [hunk 3]
```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
...
-       end do
-       _RETURN(_SUCCESS)
+        km = size(ptrin,3)
+        if (km == this%nedge -1) then
+           ks_ => this%ks
+           weights_ => this%weight
+        else
+           ks_ => this%ks_e
+           weights_ => this%weight_e
+        endif
```
Comment (tclune) [inline base/MAPL_VerticalMethods.F90:352]: This logic suggests to me that the code further down should be a helper procedure.  The conditional would then either pass (ks,weight) or (ks_e, weight_e) to that helper procedure.    But possibly there are other details that would make this less elegannt.

File: base/MAPL_VerticalMethods.F90 [hunk 4]
```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
...
+                    this%weight(:,:,lev) = (pb-pp)/(pb-pt)
+                 end where
+                 pb = pt
+               enddo
+            enddo
+            deallocate(this%pl3d) ! not needed any more. release the memory
+
+            ! for edge values
+            if(allocated(this%ks_e))  deallocate(this%ks_e)
+            if(allocated(this%weight_e)) deallocate(this%weight_e)
```
Comment (tclune) [inline base/MAPL_VerticalMethods.F90:302]: This is a bit unusual.   Do you have a situation where the object is initialized a 2nd time?  As opposed to a new object?

File: base/MAPL_VerticalMethods.F90 [hunk 5]
```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
...
+            this%nedge = km_e
+            allocate(ple3d(D1,D2,km_e))
+            ple3d = this%ple3d 
+            do lev =1, levo
+               pp = flip_sign*this%interp_levels(lev)
+               pb = flip_sign*ple3d(:,:,km_e)
+               !where (pp > pb)
+               !   this%ks_e(:,:,lev) = km_e
+               !   this%weight_e(:,:,lev) = 1.
+               !endwhere
```
Comment (tclune) [inline base/MAPL_VerticalMethods.F90:315]: Delete commented code?
```suggestion
```

File: base/MAPL_VerticalMethods.F90 [hunk 6]
```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
...
+        ptrout = MAPL_UNDEF
+        do lev = 1, levo
+          do j = 1, D2
+            do i = 1, D1
+               k = ks_(i,j,lev)
+               if (k == -1) cycle
+               !if (k == km) then
+               !   ptrout(i,j,lev) = ptrin(i,j,k)
+               !   cycle
+               !endif
```
Comment (tclune) [inline base/MAPL_VerticalMethods.F90:366]: Fixing previous suggestion.
```suggestion
```

File: base/MAPL_VerticalMethods.F90 (hunk 4 above)
Comment (weiyuan-jiang) [reply base/MAPL_VerticalMethods.F90:302]: I actually am not sure if this object is reused in different collections which may have different dimensions. This is just to be on the safe side.  The other arrays  (pl3d, ple3d) are reallocated, so I follow that logic.

File: base/MAPL_VerticalMethods.F90 (hunk 3 above)
Comment (weiyuan-jiang) [reply base/MAPL_VerticalMethods.F90:348]: The code further down just uses  ks_ and weighs_.  They don't know if it is center or edge any more.

**Review by tclune — DISMISSED** [review]

**Review by mathomp4 — APPROVED** [review]
Seems to be zero-diff.

**Pull request merged**